File: C++/Solution56.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
using namespace std;

class Solution56 {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> res;
        int left = intervals[0][0], right = intervals[0][1];
        for(int index = 1; index < intervals.size(); index++){
            if(intervals[index][0] <= right){
                right = max(right, intervals[index][1]);
            }
            else{
                res.push_back({left, right});
                left = intervals[index][0];
                right = intervals[index][1];
            }
        }
        res.push_back({left, right});
        return res;
    }
};


int main(){
   Solution56 s;
    vector<vector<int>> test = {{1, 4}, {4, 5}};
    vector<vector<int>> res = s.merge(test);
    return 0;
 }

File: Java/greedy/Solution56.java
import java.util.Arrays;
import java.util.Comparator;

class Solution56 {
    public int[][] merge(int[][] intervals) {
    	Comparator<int[]> cmp = (x, y) -> x[0] - y[0];
    	Arrays.sort(intervals, cmp);
    	int[][] result = new int[intervals.length][2];
    	int index = 0;
    	int left = intervals[0][0];
    	int right = intervals[0][1];
    	for(int i = 1; i < intervals.length; i++){
    		if(intervals[i][0] > right){
    			result[index] = new int[]{left, right};
    			index += 1;
    			left = intervals[i][0];
    			right = intervals[i][1];
    		}
    		else{
    			right = Math.max(right, intervals[i][1]);
    		}
    	}
    	result[index] = new int[]{left, right};
    	index += 1;
    	int[][] result2 = new int[index][2];
    	System.arraycopy(result, 0, result2, 0, index);
    	return result2;
    }

    public static void main(String[] args){
    	Solution56 s = new Solution56();
    	int[][] test1 = new int[][]{new int[]{1, 3}, new int[]{2, 6}, new int[]{8, 10}, new int[]{15, 18}};
    	int[][] result1 = s.merge(test1);
    }
}